File: src/main/java/composing/methods/StockReportGenerator.java
package composing.methods;

import java.util.LinkedHashMap;
import java.util.stream.Collectors;

public class StockReportGenerator {
    public String generate(Warehouse warehouse) {
        return createReport(warehouse.getId())
                .append(formatStock(warehouse.getStock()))
                .append("\nTotal: ").append(warehouse.calculateStockValue()).append("€")
                .toString();
    }

    private StringBuilder createReport(int id) {
        return new StringBuilder("Report for warehouse : " + id + "\n");
    }

    private String formatStock(LinkedHashMap<Product, Integer> stock) {
        return stock
                .entrySet()
                .stream()
                .map(entry -> "Product: " + entry.getKey().getName() + " Price: " + entry.getKey().getPrice() + " Stock : " + entry.getValue() + " units")
                .collect(Collectors.joining("\n"));
    }
}

File: src/scala/src/main/scala/composing/methods/StockReportGenerator.scala
package composing.methods

import scala.util.Properties.lineSeparator

object StockReportGenerator {
  def generate: Warehouse => String =
    (warehouse: Warehouse) => {
      createReport(warehouse) +
        formatStock(warehouse) +
        s"Total: ${warehouse.calculateStockValue}€"
    }

  private def createReport(warehouse: Warehouse): String =
    s"Report for warehouse : ${warehouse.id}$lineSeparator"

  private def formatStock(warehouse: Warehouse): String = {
    warehouse.stock.map {
      case (product, count) =>
        s"Product: ${product.name} Price: ${product.price} Stock : $count units$lineSeparator"
    }.mkString
  }
}
